File: Beginner/C/String/MyAtoi.c
int myAtoi(char* str) {
    long long res = 0;
    int tag = 1;
    while (*str == ' ')
        str ++;
    if ((*str < '0' || *str > '9') && *str != '-' && *str != '+')
        return 0;
    if (*str == '-') {
        tag = -1; str ++;
    }
    else if (*str == '+')
        str ++;
    int count = 0;
    while (*str == '0')
        str ++;
    while (*str != '\0' && *str >= '0' && *str <= '9') {
        count ++;
        if (count > 10) {
            if (tag < 0)
                return INT_MIN;
            return INT_MAX;
        }
        res = *str - '0' + res * 10;
        str ++;
    }
    res *= tag;
    if (res < INT_MIN)
        return INT_MIN;
    if (res > INT_MAX)
        return INT_MAX;
    return res;
}

File: Beginner/Java/String/MyAtoi.java
class Solution {
    public int myAtoi(String str) {
        str = str.trim();
        int tag = 1, pos = 0;
        long res = 0;
        if (str.length() > 0) {
            char first = str.charAt(0);
            if (first == '-') {
                tag = -1; pos = 1;
            }
            else if (first == '+')
                pos = 1;
            while (pos < str.length() && Character.isDigit(str.charAt(pos))) {
                res = res * 10 + (str.charAt(pos ++) - '0');
                if (res * tag > Integer.MAX_VALUE)
                    return Integer.MAX_VALUE;
                else if (res * tag < Integer.MIN_VALUE)
                    return Integer.MIN_VALUE;
            }
        }
        return (int)(res * tag);
    }
}
